File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;
class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int tw = 0;
        Queue<int []> q = new LinkedList<>();
        for(int i=0;i<truck_weights.length;i++){
            int now = truck_weights[i];
            if(tw+now<=weight){
                if(!q.isEmpty()){
                    int [] temp = q.peek();
                    if(temp[1]+(bridge_length-1)==answer){
                        tw-=temp[0];
                        q.poll();
                    }
                }
                answer++;
                tw += now;
                q.add(new int[] {now, answer});
            }else{
                int [] temp = q.poll();
                answer=temp[1]+bridge_length-1;
                tw -= temp[0];
                i--;
            }
            //System.out.println(answer+" "+q.size());
        }
        return answer+bridge_length;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    var answer = 0;
    let w = 0;
    const bridge = [];
    for(let i=0;i<truck_weights.length;i++){
        const now = truck_weights[i]
        while(w+now>weight){
            const [first_truck,time] = bridge.shift();
            answer = Math.max(answer,time+bridge_length)
            w-=first_truck
        }
        bridge.push([now,answer])
        w+=now
        answer++
    }
    if(bridge.length>0){
        answer = Math.max(answer,bridge[bridge.length-1][1]+bridge_length+1)
    }
    return answer;
}